osmo_jupyter/db_access.py

```diff
@@ -0,0 +1,50 @@
+
+def connect_to_db():
+    # TODO
+
+def load_calculation_details(db_engine, node_ids, start_utc, end_utc, include_hub_id=False, downsample_interval=None):
+    """ Load node data from the calculation_details table, optionally with hub ID included from the readings table
+
+    Args:
+        db_engine: database engine created using `connect_to_db`
+        node_ids: iterable of node IDs to get data for
+        start_utc: string of ISO-formatted start datetime in UTC, inclusive # TODO (PR feedback plz): establish safe patterns for use of local time and use that instead
+        end_utc: string of ISO-formatted end datetime in UTC, inclusive
+        include_hub_id: if True, the output will include a 'hub_id' column. Default False because the request including hub_id takes extra time.
+        downsample_interval: if this is a number, it will be used to select fewer rows. You should get *roughly* n / downsample_interval samples.
+    Returns:
+        a pandas.DataFrame of data from the node IDs provided.
+    Raises:
+        sqlalchemy.OperationalError: database connection is not working (often due to wrong password or network disconnect). In this case, a good debugging step is to reconnect to the database.
+    """
+
+    connection = db_engine.connect()
+
+    # TODO? (PR feedback plz): it feels like there's already a fair amount of logic in constructing this query; I could factor out a construct_node_data_query() function here for unit testing
+    downsample_clause = f'AND MOD(calculation_detail.reading_id, {downsample_interval}) = 0' if downsample_interval is not None else ''

```

Comment:
hmm, sounds like the ORM that Jeremy was suggesting...
someday?

Reply:
hmm, in my head it was much simpler than `ORM` suggests - more like what you suggested here, but taking it in a slightly different direction to limit the number of random parameters going into `load_calculation_details`. something like:

```
query = construct_node_data_query(node_ids= ..., include_hub_id=.., <other random parameters that get shimmed in here>)

data = load_calculation_details(db_engine= ..., query=query)
```

Maybe that's not helping that much since it's just moving the messiness to a different function, but it feels slightly cleaner to me.

Reply:
yeah, I see the draw of what you're saying, but I'd like to keep the number of necessary function calls to a minimum here (since the objective of the library is to simplify things for users).

Reply:
yeah I was kind of having that thought too. I guess if you wanted to spend a little more time thinking about how to factor this in a way that helps pave a good path forward for people tweaking it while still maintaining it's ease-of-use, I would be in favor of that. but if not, I suppose I'm okay with it this way too.